datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -710,6 +719,82 @@ impl RecordBatchStream for SMJStream {
     }
 }
 
+#[inline(always)]
+fn last_index_for_row(
+    row_index: usize,
+    indices: &UInt64Array,
+    ids: &[usize],
+    indices_len: usize,
+) -> bool {
+    row_index == indices_len - 1
+        || ids[row_index] != ids[row_index + 1]
+        || indices.value(row_index) != indices.value(row_index + 1)
+}
+
+// Returns a corrected boolean bitmask for the given join type
+// Values in the corrected bitmask can be: true, false, null
+// `true` - the row found its match and sent to the output
+// `null` - the row ignored, no output
+// `false` - the row sent as NULL joined row
+fn get_corrected_filter_mask(
+    join_type: JoinType,
+    indices: &UInt64Array,
+    ids: &[usize],
+    filter_mask: &BooleanArray,
+    expected_size: usize,
+) -> Option<BooleanArray> {
+    let streamed_indices_length = indices.len();
+    let mut corrected_mask: BooleanBuilder =
+        BooleanBuilder::with_capacity(streamed_indices_length);
+    let mut seen_true = false;
+
+    match join_type {
+        JoinType::Left => {
+            for i in 0..streamed_indices_length {
+                let last_index =
+                    last_index_for_row(i, indices, ids, streamed_indices_length);
+                if filter_mask.value(i) {
+                    seen_true = true;
+                    corrected_mask.append_value(true);
+                } else if seen_true || !filter_mask.value(i) && !last_index {
+                    corrected_mask.append_null(); // to be ignored and not set to output
+                } else {
+                    corrected_mask.append_value(false); // to be converted to null joined row
+                }
+
+                if last_index {
+                    seen_true = false;
+                }
+            }
+
+            // Generate null joined rows for records which have no matching join key
+            let null_matched = expected_size - corrected_mask.len();
+            corrected_mask.extend(vec![Some(false); null_matched]);
+            Some(corrected_mask.finish())
+        }
+        JoinType::LeftSemi => {
+            for i in 0..streamed_indices_length {
+                let last_index =
+                    last_index_for_row(i, indices, ids, streamed_indices_length);
+                if filter_mask.value(i) && !seen_true {
+                    seen_true = true;
+                    corrected_mask.append_value(true);
+                } else {
+                    corrected_mask.append_null(); // to be ignored and not set to output
+                }
+
+                if last_index {
+                    seen_true = false;
+                }
+            }
+
+            Some(corrected_mask.finish())
+        }
+
+        _ => None,
```

Comment:
I think it would be good to list out the possibilities here -- I think it is only `JoinType::Inner` right?

Reply:
correct, perhaps Full Join won't also need a separate mask as Full Join is a specific case of Left and Right.